_posts/2019-09-19-dask-hyperparam-opt.md

```diff
@@ -0,0 +1,440 @@
+
+First, let's look at the parameters required for Dask-ML's implementation
+of Hyperband (which is in the class `HyperbandSearchCV`).
+
+### Hyperband parameters: rule-of-thumb
+`HyperbandSearchCV` has two inputs:
+
+1. `max_iter`, which determines how many times to call `partial_fit`
+2. the chunk size of the Dask array, which determines how many data each
+   `partial_fit` call receives.
+
+These fall out pretty naturally once it's known how long to train the best
+model and very approximately how many parameters to sample:
+
+``` python
+n_examples = 50 * len(X_train)  # 50 passes through dataset for best model
+n_params = 299  # sample about 300 parameters
+
+# inputs to hyperband
+max_iter = n_params
+chunk_size = n_examples // n_params
+```
+
+This rule-of-thumb requires are exactly the inputs the user cares about:
+
+* `n_params`, a measure of how complex the search space is
+* `n_examples`, how long to train the best model.
+
+Notably, there's no tradeoff between `n_examples` and `n_params` like with
+Scikit-learn's `RandomizedSearchCV` because `n_examples` is only for *some*
+models, not for *all* models. There's more details on this
+rule-of-thumb in the "Notes" section of the [`HyperbandSearchCV`
+docs][hyperband-docstring].
+
+[hyperband-docstring]:https://ml.dask.org/modules/generated/dask_ml.model_selection.HyperbandSearchCV.html#dask_ml.model_selection.HyperbandSearchCV
+
+With these, a `HyperbandSearchCV` object can easily be created.
+
+### Finding the best performing hyperparameters
+
+This model selection algorithm Hyperband is implemented in the class
+`HyperbandSearchCV`. Let's create an instance of that class:
+
+``` python
+>>> from dask_ml.model_selection import HyperbandSearchCV
+>>>
+>>> search = HyperbandSearchCV(
+...     est, params, max_iter=max_iter, aggressiveness=4
+... )
+```
+
+`aggressiveness` defaults to 3. `aggressiveness=4` is chosen because this is an
+*initial* search; I know nothing about how this search space. Then, this search
+should be more aggressive in culling off bad models.
+
+I may be worried about how long the computation will take, or how many models
+are created:
+
+``` python
+>>> search.metadata["n_models"]
+378
+>>> search.metadata["partial_fit_calls"]
+5721
+```
+
+Now that we have some idea on how long the computation will take, let's ask it
+to find the best set of hyperparameters:
+
+``` python
+>>> from dask_ml.model_selection import train_test_split
+>>> X_train, y_train, X_test, y_test = train_test_split(X, y)
+>>>
+>>> X_train = X_train.rechunk(chunk_size)
+>>> y_train = y_train.rechunk(chunk_size)
+>>>
+>>> search.fit(X_train, y_train)
+```
+
+The dashboard will be active during this time[^dashboard]:
+
+[^dashboard]:But it probably won't be this fast: the video is sped up by a factor of 3.
+
+<p>
+<video width="600" autoplay loop controls class="center">
+  <source src="/images/2019-hyperband/dashboard-compress.mp4" type="video/mp4" class="center">
```

Comment:
Do we have other videos in this repository? Or can we upload it to youtube and embed the video?

Reply:
Currently, I only have that one video (which depends on https://github.com/dask/dask-ml/pull/528).

I'm not opposed to making another video.

Reply:
Given that the video is so small, it's probably fine to include it here in the git repostitory.

Reply:
Ah, gotcha. I only include a compressed version in git (I think 700kB). The uncompressed version is too large for git, 27 MB.